File: packages/py/src/config_discovery/sources.py
"""The ``sources`` output - SPEC section 7.

This is the only thing standing between a user who expected ``debug`` and an afternoon of
guessing which of six layers set ``trace``, so every source that was read is listed, winners and
losers alike.
"""

from __future__ import annotations

from pathlib import Path, PurePosixPath

from .types import Layer, Source


def _rewrite(path: str, relative_to: Path | None) -> str:
    # <defaults>, <env> and <overrides> are labels, not paths, and are passed through.
    candidate = Path(path)
    if relative_to is None or not candidate.is_absolute():
        return path
    try:
        relative = candidate.relative_to(Path(relative_to))
    except ValueError:
        # Outside the fixture root: stay absolute rather than emit a ../../ climb that no
        # expected.json could match.
        return PurePosixPath(candidate).as_posix()
    return PurePosixPath(relative).as_posix()


def build_sources(layers: list[Layer], *, relative_to: Path | None = None) -> list[Source]:
    """Entry order is **application order** - the order the layers were merged, lowest effective
    priority first (SPEC section 3.1).

    That is ascending precedence with one documented exception: a root's ``.env`` (precedence 3)
    belongs inside that root's block, so a user-level ``.env`` still loses to a project-local
    ``config.toml``. Sorting this list by precedence would reorder it into something that does
    not describe what happened.
    """
    return [
        Source(
            path=_rewrite(layer.source.path, relative_to),
            format=layer.source.format,
            precedence=layer.source.precedence,
            keys=sorted(layer.value.keys()) if isinstance(layer.value, dict) else [],
        )
        for layer in layers
    ]

File: packages/java/core/src/main/java/dev/configdiscovery/Sources.java
package dev.configdiscovery;

import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

/**
 * The {@code sources} output - SPEC section 7.
 *
 * <p>The only thing standing between a user who expected {@code debug} and an afternoon of
 * guessing which of six layers set {@code trace}, so every source that was read is listed,
 * winners and losers alike.
 */
public final class Sources {

    private Sources() {}

    /**
     * Emit entries in <strong>application order</strong> - the order the layers were merged,
     * lowest effective priority first (SPEC section 3.1).
     *
     * <p>That is ascending precedence with one documented exception: a root's {@code .env}
     * (precedence 3) belongs inside that root's block, so a user-level {@code .env} still loses to
     * a project-local {@code config.toml}. Sorting this list by precedence would reorder it into
     * something that does not describe what happened.
     */
    public static List<Source> build(List<Merge.Layer> layers, Path relativeTo) {
        List<Source> sources = new ArrayList<>(layers.size());
        for (Merge.Layer layer : layers) {
            List<String> keys = new ArrayList<>();
            layer.value().fieldNames().forEachRemaining(keys::add);
            keys.sort(String::compareTo);
            sources.add(new Source(
                    rewrite(layer.source().path(), relativeTo),
                    layer.source().format(),
                    layer.source().precedence(),
                    keys));
        }
        return List.copyOf(sources);
    }

    static String rewrite(String path, Path relativeTo) {
        // <defaults>, <env> and <overrides> are labels, not paths, and are passed through.
        Path candidate = Path.of(path);
        if (relativeTo == null || !candidate.isAbsolute()) {
            return path;
        }
        if (!candidate.startsWith(relativeTo)) {
            // Outside the fixture root: stay absolute rather than emit a ../../ climb that no
            // expected.json could match.
            return slash(candidate);
        }
        return slash(relativeTo.relativize(candidate));
    }

    private static String slash(Path path) {
        // Forward slashes on every platform - on Windows this is not a no-op.
        StringBuilder text = new StringBuilder();
        for (Path segment : path) {
            if (text.length() > 0) {
                text.append('/');
            }
            text.append(segment);
        }
        String result = text.toString();
        return path.isAbsolute() ? "/" + result : result;
    }
}
